server/parsec/realm_export.py

```diff
@@ -0,0 +1,1174 @@
+        self._queries_queue.put_nowait((cb, ret_queue))
+        match await ret_queue.get():
+            case Exception() as exc:
+                raise exc
+            case ret:
+                return ret
+
+
+type ToExportBytes = int  # Total amount of data in bytes to be exported
+type ExportedBytes = int  # Amount of data in bytes that have been exported so far
+type ExportProgressStep = (
+    Literal["certificates_start"]
+    | Literal["certificates_done"]
+    | tuple[Literal["vlobs"], ExportedBytes, ToExportBytes]
+    | tuple[Literal["blocks_metadata"], ExportedBytes, ToExportBytes]
+    | tuple[Literal["blocks_data"], ExportedBytes, ToExportBytes]
+)
+type ProgressReportCallback = Callable[[ExportProgressStep], None]
+
+
+def get_earliest_allowed_snapshot_timestamp() -> DateTime:
+    """
+    The very first step we do when exporting a realm is to determine what should be
+    exported according to the snapshot timestamp parameter.
+
+    Obviously snapshot timestamp is not allowed to be beyond present time since only
+    Bardock can see in the future.
+
+    On top of that, it is also bad to have the snapshot timestamp in the past but
+    too close to the present time as the server is still allowed accept data timed
+    before our snapshot timestamp.
+    """
+    return DateTime.now().subtract(seconds=BALLPARK_CLIENT_LATE_OFFSET)
+
+
+async def export_realm(
+    backend: Backend,
+    organization_id: OrganizationID,
+    realm_id: VlobID,
+    snapshot_timestamp: DateTime,
+    output_db_path: Path,
+    on_progress: ProgressReportCallback,
+):
+    earlier_allowed_timestamp = get_earliest_allowed_snapshot_timestamp()
+    if snapshot_timestamp > earlier_allowed_timestamp:
+        raise RealmExporterInputError(
+            f"Snapshot timestamp cannot be more recent than present time - {BALLPARK_CLIENT_LATE_OFFSET}s (i.e. `{earlier_allowed_timestamp}`)"
+        )
+
+    # 0) Ensure the export is possible since organization & realm exist on the server
+
+    outcome = await backend.realm.export_do_base_info(
+        organization_id=organization_id, realm_id=realm_id, snapshot_timestamp=snapshot_timestamp
+    )
+    match outcome:
+        case RealmExportDoBaseInfo() as base_info:
+            pass
+        case RealmExportDoBaseInfoBadOutcome.ORGANIZATION_NOT_FOUND:
+            raise RealmExporterInputError(f"Organization `{organization_id.str}` doesn't exists")
+        case RealmExportDoBaseInfoBadOutcome.REALM_NOT_FOUND:
+            raise RealmExporterInputError(
+                f"Realm `{realm_id.hex}` doesn't exist in organization `{organization_id.str}`"
+            )
+        case RealmExportDoBaseInfoBadOutcome.REALM_DIDNT_EXIST_AT_SNAPSHOT_TIMESTAMP:
+            raise RealmExporterInputError(
+                f"Requested snapshot timestamp `{snapshot_timestamp}` is older than realm creation"
+            )
+
+    # 1) Create the output SQLite database (or re-open if it already exists)
```

Comment:
It might be useful to have [your answer](https://github.com/Scille/parsec-cloud/pull/9342/files#r1933473533) in the comments even if a little verbose:
```suggestion
    # 1) Create the output SQLite database (or re-open if it already exists).
    #
    #    Note that the database stores everything needed to determine if the export
    #    is complete, and if not, everything to be able to continue it ;-)
    #
    #    So the following commands lead to the same foo.sqlite (even if subsequent
    #    export occurs after additional changes are added to the export):
    #    $ realm_export ./foo.sqlite 
    #    $ realm_export ./foo.sqlite &; sleep 1; kill %1; realm_export ./foo.sqlite 
```

Reply:
:+1: